fix karabiner-elements on macos

auscyber opened the pull request:
its a bit of a crazy nightmare and depends on my other pr #1594 but it does work finally responding finally to #1041

auscyber commented:
if needed, i could split up the driverkit stuff into a separate option so that it can be used with kanata and kmonad

Enzime commented on modules/services/karabiner-elements/default.nix:
```diff
@@ -1,28 +1,46 @@
 
 let
   cfg = config.services.karabiner-elements;
+  package = cfg.package.overrideAttrs (attrs: {
+    fixupPhase = ''
+      find $out -type f -name '._embedded.provisionprofile' -exec rm -rf {} \;
+      find $driver -type f -name '._embedded.provisionprofile' -exec rm -rf {} \;
+    '';
+    dontPatch = true;
+  });
```
This should be upstreamed into Nixpkgs

Enzime commented on modules/services/karabiner-elements/default.nix:
```diff
@@ -91,26 +102,35 @@ in
-    environment.userLaunchAgents."org.pqrs.karabiner.agent.karabiner_observer.plist".source = "${cfg.package}/Library/LaunchAgents/org.pqrs.karabiner.agent.karabiner_observer.plist";
-    environment.userLaunchAgents."org.pqrs.karabiner.karabiner_console_user_server.plist".source = "${cfg.package}/Library/LaunchAgents/org.pqrs.karabiner.karabiner_console_user_server.plist";
+    # from launch agents folder in karabiner elements executable
+    environment.userLaunchAgents."org.pqrs.service.agent.karabiner_grabber.plist".source =
+      "/Library/Application Support/org.pqrs/Karabiner-Elements/Karabiner-Elements Non-Privileged Agents.app/Contents/Library/LaunchAgents/org.pqrs.service.agent.karabiner_grabber.plist";
+    environment.userLaunchAgents."org.pqrs.service.agent.karabiner_console_user_server.plist".source =
+      "/Library/Application Support/org.pqrs/Karabiner-Elements/Karabiner-Elements Non-Privileged Agents.app/Contents/Library/LaunchAgents/org.pqrs.service.agent.karabiner_console_user_server.plist";
+    environment.userLaunchAgents."org.pqrs.service.agent.Karabiner-Menu.plist".source =
+      "/Library/Application Support/org.pqrs/Karabiner-Elements/Karabiner-Elements Non-Privileged Agents.app/Contents/Library/LaunchAgents/org.pqrs.service.agent.Karabiner-Menu.plist";
+    environment.userLaunchAgents."org.pqrs.service.agent.Karabiner-MultitouchExtension.plist".source =
+      "/Library/Application Support/org.pqrs/Karabiner-Elements/Karabiner-Elements Non-Privileged Agents.app/Contents/Library/LaunchAgents/org.pqrs.service.agent.Karabiner-MultitouchExtension.plist";
+    environment.userLaunchAgents."org.pqrs.service.agent.Karabiner-NotificationWindow.plist".source =
+      "/Library/Application Support/org.pqrs/Karabiner-Elements/Karabiner-Elements Non-Privileged Agents.app/Contents/Library/LaunchAgents/org.pqrs.service.agent.Karabiner-NotificationWindow.plist";
+
```
I think these should come from inside the package instead and shouldn't require linking `/Library/Application Support`

We already do some level of patching in Nixpkgs to try to make these work https://github.com/NixOS/nixpkgs/commit/cfe3c23ad5c61c6151ccb0879a467391121f0553

Enzime commented on modules/services/karabiner-elements/default.nix:
```diff
@@ -1,28 +1,46 @@
   options.services.karabiner-elements = {
     enable = mkEnableOption "Karabiner-Elements";
     package = mkPackageOption pkgs "karabiner-elements" { };
+    overrideFixup = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        Whether to override the default fixup phase to make Karabiner-Elements work
+        correctly when installed outside of /Applications. This is necessary if you
+        are using a custom location for Nix Apps such as /Applications/Nix Apps.
+      '';
+    };
```
I think this option is unused?

daneov commented:
With that latest one the flake applies and installs the package, but I keep getting these for both daemons :

<img width="242" height="245" alt="image" src="https://github.com/user-attachments/assets/7a29be45-fef1-4d81-a8dd-5ebbf9a2d1a6" />

I'm on macOS Tahoe, haven't tried the official install to see if that makes a difference. Should I, or does this ring a bell for you in relation to the PR?

daneov commented:
@Enzime: is this PR an appropriate solution for #1041?

auscyber commented on modules/services/karabiner-dk/default.nix:
```diff
@@ -0,0 +1,51 @@
+        "activate"
+      ];
+      serviceConfig.RunAtLoad = true;
+      managedBy = "auscybernix.keybinds.karabiner-driver-kit.enable";
```
yeah oops

daneov commented:
As an enthusiast it'd be nice to get Karabiner working. Would this PR be an appropriate solution to that, @Enzime?

Not meaning to be pushy, but I'd also prefer to just depend on nix-darwin and not on a fork to get this functionality :)
Thanks!

AdoPi commented:
Yes it was authorized to run in my case, at least from what I remember.

auscyber committed:
karabiner-elements: fix on macos

karabiner-dk: add module

system: add application support linking using new option
`environment.extraAppPaths`

auscyber commented:
i've fixed and changed this to use the new pathsToLink and this is a lot less sus when it comes to copying folders into /Library

yshym commented on modules/services/karabiner-dk/default.nix:
```diff
@@ -0,0 +1,55 @@
+let
+  cfg = config.services.karabiner-dk;
+
+  parentAppDir = "/Applications/Nix Apps";
```
We should add a way to reference this instead of redefining otherwise we'll have to change the directory path in multiple places

yshym commented on modules/services/karabiner-dk/default.nix:
```diff
@@ -0,0 +1,55 @@
+      ];
+      serviceConfig.RunAtLoad = true;
+      managedBy = "services.karabiner-dk.enable";
+    };
```
Why do you need this user agent if you already execute the same command in `launchd.daemons.start-karabiner-dk`?

yshym commented on modules/services/karabiner-dk/default.nix:
```diff
@@ -0,0 +1,55 @@
+        		spctl -a -vvv -t install "${parentAppDir}/.Karabiner-VirtualHIDDevice-Manager.app/Contents/MacOS/Karabiner-VirtualHIDDevice-Manager"
+                "${parentAppDir}/.Karabiner-VirtualHIDDevice-Manager.app/Contents/MacOS/Karabiner-VirtualHIDDevice-Manager" activate
+      '';
+      serviceConfig.Label = "org.nixos.start-karabiner-dk";
```
You may remove this line as it is the default value

yshym commented on modules/services/karabiner-dk/default.nix:
```diff
@@ -0,0 +1,55 @@
+    ];
+    environment.extraAppPaths = [
+      "/Library/Application Support/org.pqrs/Karabiner-DriverKit-VirtualHIDDevice"
+    ];
```
What is the use case when a person which uses this nix module (which already manages the daemon automatically) would run `/Library/Application Support/org.pqrs/Karabiner-DriverKit-VirtualHIDDevice/Applications/Karabiner-VirtualHIDDevice-Daemon.app/Contents/MacOS/Karabiner-VirtualHIDDevice-Daemon` manually? We don't need to change neither `system.build.applications` nor `system.activationScripts.applications` if there isn't any

yshym commented on modules/services/karabiner-dk/default.nix:
```diff
@@ -0,0 +1,55 @@
+    launchd.daemons.Karabiner-DriverKit-VirtualHIDDevice-Daemon = {
+      serviceConfig.ProgramArguments = [
+        "/Library/Application Support/org.pqrs/Karabiner-DriverKit-VirtualHIDDevice/Applications/Karabiner-VirtualHIDDevice-Daemon.app/Contents/MacOS/Karabiner-VirtualHIDDevice-Daemon"
+      ];
```
You don't have to copy anything into `/Library` to run here
```suggestion
      command = "${cfg.package}/Library/Application\\ Support/org.pqrs/Karabiner-DriverKit-VirtualHIDDevice/Applications/Karabiner-VirtualHIDDevice-Daemon.app/Contents/MacOS/Karabiner-VirtualHIDDevice-Daemon";
```
